python/tests/integration_tests/wrappers/test_openai.py

```diff
@@ -180,3 +184,180 @@ async def test_completions_async_api(mock_session: mock.MagicMock, stream: bool)
+            "messages": [{"role": "user", "content": "howdy"}],
+            "stream": True,
+        },
+        "expect_usage_metadata": False,
+    },
+    {
+        "description": "non-stream usage",
+        "params": {
+            "model": "gpt-4o-mini",
+            "messages": [{"role": "user", "content": "howdy"}],
+        },
+        "expect_usage_metadata": True,
+    },
+    {
+        "description": "complex usage",
+        "params": {
+            "model": "o1-mini",
+            "messages": [
+                {
+                    "role": "user",
+                    "content": (
+                        "Write a bash script that takes a matrix represented "
+                        "as a string with format '[1,2],[3,4],[5,6]' and prints the "
+                        "transpose in the same format."
+                    ),
+                }
+            ],
+        },
+        "expect_usage_metadata": True,
+        "check_reasoning_tokens": True,
+    },
+]
+
+
+@pytest.mark.parametrize("test_case", test_cases)
+@mock.patch("langsmith.client.requests.Session")
+def test_wrap_openai_chat(mock_session: mock.MagicMock, test_case):
+    import openai
+    from openai.types.chat import ChatCompletion, ChatCompletionChunk
+
+    oai_client = openai.Client()
+    ls_client = langsmith.Client(session=mock_session())
+    wrapped_oai_client = wrap_openai(oai_client, tracing_extra={"client": ls_client})
+
+    collect = Collect()
+    run_id_to_usage_metadata = {}
+    with langsmith.tracing_context(enabled=True):
+        params: dict[str, Any] = test_case["params"].copy()
+        params["langsmith_extra"] = {"on_end": collect}
+        res = wrapped_oai_client.chat.completions.create(**params)
+
+        if params.get("stream"):
+            for chunk in res:
+                assert isinstance(chunk, ChatCompletionChunk)
+                if test_case.get("expect_usage_metadata") and hasattr(chunk, "usage"):
+                    oai_usage = chunk.usage
+        else:
+            assert isinstance(res, ChatCompletion)
+            oai_usage = res.usage
+
+        if test_case["expect_usage_metadata"]:
+            usage_metadata = collect.run.outputs["usage_metadata"]
+            assert usage_metadata["input_tokens"] == oai_usage.prompt_tokens
+            assert usage_metadata["output_tokens"] == oai_usage.completion_tokens
+            assert usage_metadata["total_tokens"] == oai_usage.total_tokens
+            if test_case.get("check_reasoning_tokens"):
+                assert (
+                    usage_metadata["output_token_details"]["reasoning"]
+                    == oai_usage.completion_tokens_details.reasoning_tokens
+                )
+        else:
+            assert collect.run.outputs.get("usage_metadata") is None
+            assert collect.run.outputs.get("usage") is None
+
+        run_id_to_usage_metadata[collect.run.id] = collect.run.outputs.get(
+            "usage_metadata"
+        )
+
+    time.sleep(0.1)
```

Comment:
@hinthornw is there a better way to give the thread a chance besides sleeping? I feel like we had some wait for all tracers functionality back in the day

Reply:
Not really. I do a `_wait_for` function that loops; or you could not do autobatching